pipeline/mutate/mutator_hydrator_test.go

```diff
@@ -346,27 +406,85 @@ func TestMutatorHydrator(t *testing.T) {
 				Match:   newAuthenticationSession(),
 				Err:     nil,
 			},
+			cacheTestCustomCacheKeyName: {
+				Setup:   routerAuthSessionCache(),
+				Session: newAuthenticationSession(setSubject(sampleSubject)),
+				Rule:    &rule.Rule{ID: "test-rule"},
+				Config:  configWithSpecialCacheKey(sampleSubject),
+				Request: &http.Request{},
+				Match:   newAuthenticationSession(setSubject(sampleSubject)),
+				Err:     nil,
+			},
+			cacheTestAuthSessionName: {
+				Setup:   routerAuthSessionCache(),
+				Session: newAuthenticationSession(setSubject(sampleSubject)),
+				Rule:    &rule.Rule{ID: "test-rule"},
+				// An empty cache key will ensure default (AuthenticationSession) behavior is applied.
+				Config:  configWithSpecialCacheKey(""),
+				Request: &http.Request{},
+				Match:   newAuthenticationSession(setSubject(sampleSubject)),
+				Err:     nil,
+			},
+			cacheTestCustomKeyTemplateName: {
+				Setup:   routerAuthSessionCache(),
+				Session: newAuthenticationSession(setSubject(cacheTestTemplateSubject)),
+				Rule:    &rule.Rule{ID: "test-rule"},
+				Config:  configWithSpecialCacheKey("{{ print .Subject }}"),
+				Request: &http.Request{},
+				Match:   newAuthenticationSession(setSubject(cacheTestTemplateSubject)),
+				Err:     nil,
+			},
 		}
 
 		for testName, specs := range testMap {
 			t.Run(testName, func(t *testing.T) {
 				var router http.Handler
 				var ts *httptest.Server
+
 				if specs.Setup != nil {
 					router = specs.Setup(t)
 				}
 				ts = httptest.NewServer(router)
 				defer ts.Close()
 
-				err := a.Mutate(specs.Request, specs.Session, specs.Config(ts), specs.Rule)
-				if specs.Err == nil {
+				switch {
+				case testName == cacheTestCustomKeyTemplateName:
+					fallthrough
+				case testName == cacheTestCustomCacheKeyName:
+					specs.Session.Extra = make(map[string]interface{})
+					specs.Session.Extra[cacheTestKeyName] = struct{}{}
+					require.NoError(t, a.Mutate(specs.Request, specs.Session, specs.Config(ts), specs.Rule))
+					// Delete K/V-combination above. Must be served from the cache,
+					// K/V-combination present ensure session originates from cache.
+					delete(specs.Session.Extra, cacheTestKeyName)
+					// Cache entry is being written asynchronously. Obviously this here is not
+					// a good strategy, however, the alternative would be to replace the cache.
+					// See https://github.com/dgraph-io/ristretto/blob/9d4946d9b973c8e860ae42944e07f5bbe28a506b/cache_test.go#L17
+					time.Sleep(cacheTestWriteSleep)
+				case testName == cacheTestAuthSessionName:
+					// Ensure session is persisted within cache.
+					require.NoError(t, a.Mutate(specs.Request, specs.Session, specs.Config(ts), specs.Rule))
+					time.Sleep(cacheTestWriteSleep)
+				}
+
+				if err := a.Mutate(specs.Request, specs.Session, specs.Config(ts), specs.Rule); specs.Err == nil {
```

Comment:
These tests look much better! Another option you have would be to add setup and assert functions to the `testMap` and then run them here if they are not nil. That makes the test setup easier to read and follow. An alternative is also to write a dedicated test for this, you don't necessarily have to reuse the code that exists. Sometimes explicit code is better, even if there is a bit of duplication involved :)

Reply:
Done.